src/lib/sparse_ledger_lib/sparse_ledger.ml

```diff
@@ -173,6 +184,80 @@ end = struct
     ; indexes = (account_id, index) :: t.indexes
     }
 
+  let add_wide_path_unsafe tree0 path0 account =
+    (* Traverses the tree along path, collecting node and untraversed sibling hashes
+        Stops when encounters `Hash` or `Account` node.
+
+       Returns the last visited node (`Hash` or `Account`), remainder of path and
+       collected node/sibling hashes in bottom-to-top order.
+    *)
+    let rec traverse_through_nodes acc = function
+      | Tree.Hash h, rest ->
+          (acc, `Hash h, rest)
+      | Account a, rest ->
+          (acc, `Account a, rest)
+      | Node (h, l, r), `Left _ :: rest ->
+          traverse_through_nodes ((h, `Right_preserved r) :: acc) (l, rest)
+      | Node (h, l, r), `Right _ :: rest ->
+          traverse_through_nodes ((h, `Left_preserved l) :: acc) (r, rest)
+      | Node _, [] ->
+          failwith "path is shorter than a tree's branch"
+    in
+    (* Takes a list of collected node/sibling hashes in bottom-to-top order
+       (returned by the function above) and returns a tree that is reconstructed from
+       bottom to top, substituting siblings as necessary. *)
+    let build_to_top bottom_node =
+      List.fold ~init:bottom_node ~f:(fun node -> function
+        | h, `Left_preserved l ->
+            Tree.Node (h, l, node)
+        | h, `Right_preserved r ->
+            Tree.Node (h, node, r) )
+    in
+    (* Uses wide path to build the tail of path *)
+    let build_tail hash_node_to_bottom_path =
+      let bottom_el, bottom_to_hash_node_path =
+        Mina_stdlib.Nonempty_list.(rev hash_node_to_bottom_path |> uncons)
+      in
+      (* Left and right branches of a node that is parent of the bottom node *)
+      let init =
+        match bottom_el with
+        | `Left (_, h_r) ->
+            (Tree.Account account, Tree.Hash h_r)
+        | `Right (h_l, _) ->
+            (Hash h_l, Account account)
+      in
+      let f (prev_l, prev_r) = function
+        | `Left (h_l, h_r) ->
+            (Tree.Node (h_l, prev_l, prev_r), Tree.Hash h_r)
+        | `Right (h_l, h_r) ->
+            (Hash h_l, Node (h_r, prev_l, prev_r))
+      in
+      List.fold ~init bottom_to_hash_node_path ~f
+    in
+    let hash_to_top_path, hash_node, rest =
+      traverse_through_nodes [] (tree0, List.rev path0)
+    in
+    match (hash_node, Mina_stdlib.Nonempty_list.of_list_opt rest) with
+    | `Account _, None ->
+        tree0
+    | `Hash _, None ->
+        build_to_top (Tree.Account account) hash_to_top_path
+    | `Hash h, Some hash_to_bottom_path ->
+        let tail_l, tail_r = build_tail hash_to_bottom_path in
+        build_to_top (Tree.Node (h, tail_l, tail_r)) hash_to_top_path
+    | `Account _, Some _ ->
+        failwith "path is longer than a tree's branch"
+
+  let add_wide_path_unsafe (t : t) path account_id account =
+    let index =
+      List.foldi path ~init:0 ~f:(fun i acc x ->
```

Comment:
This surprised me a bit. I would think this would work if the path was in reverse order only, but I think the path is traversing from the root down to the leaf, in which case this seems wrong to me at first glance.